File: sdk/src/main/java/eu/darken/porter/sdk/PorterBinderWrapper.java
package eu.darken.porter.sdk;

import android.os.IBinder;
import android.os.IInterface;
import android.os.Parcel;
import android.os.RemoteException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.FileDescriptor;
import java.util.Objects;

/**
 * Wraps a binder so that every transaction on it is forwarded through the Porter server.
 *
 * <p>example:
 * <br><code>IPackageManager pm = IPackageManager.Stub.asInterface(new PorterBinderWrapper(binder));
 * <br>pm.getInstalledPackages(0, 0);</code>
 */
public class PorterBinderWrapper implements IBinder {

    private final IBinder original;

    public PorterBinderWrapper(@NonNull IBinder original) {
        this.original = Objects.requireNonNull(original);
    }

    @Override
    public boolean transact(int code, @NonNull Parcel data, @Nullable Parcel reply, int flags) throws RemoteException {
        // One connection answers for both the token and the transaction: a replacement publishing
        // between the two would have the server reject a parcel carrying the other wire's token.
        PorterWire wire = Porter.requireWire();
        Parcel newData = Parcel.obtain();
        try {
            newData.writeInterfaceToken(wire.descriptor());
            newData.writeStrongBinder(original);
            newData.writeInt(code);
            newData.writeInt(flags);
            newData.appendFrom(data, 0, data.dataSize());
            wire.transactRemote(newData, reply, 0);
        } finally {
            newData.recycle();
        }
        return true;
    }

    @Nullable
    @Override
    public String getInterfaceDescriptor() throws RemoteException {
        return original.getInterfaceDescriptor();
    }

    @Override
    public boolean pingBinder() {
        return original.pingBinder();
    }

    @Override
    public boolean isBinderAlive() {
        return original.isBinderAlive();
    }

    @Nullable
    @Override
    public IInterface queryLocalInterface(@NonNull String descriptor) {
        return null;
    }

    @Override
    public void dump(@NonNull FileDescriptor fd, @Nullable String[] args) throws RemoteException {
        original.dump(fd, args);
    }

    @Override
    public void dumpAsync(@NonNull FileDescriptor fd, @Nullable String[] args) throws RemoteException {
        original.dumpAsync(fd, args);
    }

    @Override
    public void linkToDeath(@NonNull DeathRecipient recipient, int flags) throws RemoteException {
        original.linkToDeath(recipient, flags);
    }

    @Override
    public boolean unlinkToDeath(@NonNull DeathRecipient recipient, int flags) {
        return original.unlinkToDeath(recipient, flags);
    }
}

File: sdk/src/main/java/eu/darken/porter/sdk/PorterBinderWrapper.kt
package eu.darken.porter.sdk

import android.os.IBinder
import android.os.IInterface
import android.os.Parcel
import android.os.RemoteException
import java.io.FileDescriptor

/**
 * Wraps a binder so that every transaction on it is forwarded through the server of one
 * connection. [PorterConnection.wrap] is the usual way to get one.
 */
internal class PorterBinderWrapper(
    private val connection: PorterConnection,
    private val original: IBinder,
) : IBinder {

    @Throws(RemoteException::class)
    override fun transact(code: Int, data: Parcel, reply: Parcel?, flags: Int): Boolean {
        connection.wire.forward(original, code, data, reply, flags)
        return true
    }

    @Throws(RemoteException::class)
    override fun getInterfaceDescriptor(): String? = original.interfaceDescriptor

    override fun pingBinder(): Boolean = original.pingBinder()

    override fun isBinderAlive(): Boolean = original.isBinderAlive

    override fun queryLocalInterface(descriptor: String): IInterface? = null

    @Throws(RemoteException::class)
    override fun dump(fd: FileDescriptor, args: Array<String>?) {
        original.dump(fd, args)
    }

    @Throws(RemoteException::class)
    override fun dumpAsync(fd: FileDescriptor, args: Array<String>?) {
        original.dumpAsync(fd, args)
    }

    @Throws(RemoteException::class)
    override fun linkToDeath(recipient: IBinder.DeathRecipient, flags: Int) {
        original.linkToDeath(recipient, flags)
    }

    override fun unlinkToDeath(recipient: IBinder.DeathRecipient, flags: Int): Boolean =
        original.unlinkToDeath(recipient, flags)
}
